assets/javascripts/lib/ajax.coffee

```diff
@@ -74,45 +76,13 @@ applyHeaders = (xhr, options) ->
 
   if options.dataType
     options.headers['Accept'] = MIME_TYPES[options.dataType] or options.dataType
+  return options.headers
 
-  for key, value of options.headers
-    xhr.setRequestHeader(key, value)
+onSuccess = (data, options) ->
+  options.success?.call options.context, data, options
   return
 
-onComplete = (xhr, options) ->
-  if 200 <= xhr.status < 300
-    if (response = parseResponse(xhr, options))?
-      onSuccess response, xhr, options
-    else
-      onError 'invalid', xhr, options
-  else
-    onError 'error', xhr, options
+onError = (type, error, options) ->
+  options.error?.call options.context, type, error, options
```

Comment:
Here are some usages for the error handler and progress callback:

- https://github.com/freeCodeCamp/devdocs/blob/676f4f6e84139b93d1bc993923ad99fd6a8b7083/assets/javascripts/app/db.coffee#L253-L257
- https://github.com/freeCodeCamp/devdocs/blob/14049dac0c18d1ea1ff7f439a63cf03a0d425b4d/assets/javascripts/app/update_checker.coffee#L14-L17
- https://github.com/freeCodeCamp/devdocs/blob/14049dac0c18d1ea1ff7f439a63cf03a0d425b4d/assets/javascripts/models/doc.coffee#L67-L70
- https://github.com/freeCodeCamp/devdocs/blob/14049dac0c18d1ea1ff7f439a63cf03a0d425b4d/assets/javascripts/models/doc.coffee#L114-L119